docs/lang_spec/Cadence Language Specification.md

```diff
@@ -0,0 +1,5798 @@
+    ;
+  
+  variableKind
+    : 'let'
+    | 'var'
+    ;
+  ```
+  - Components:
+    - **access**: Specifies the visibility level of the variable (e.g., pub, access(contract)). It is optional.
+    - **variableKind**: Specifies whether the declaration is a constant (let) or a variable (var).
+    - **identifier**: The name of the variable or constant.
+    - **typeAnnotation** (optional): Specifies the type of the variable. If omitted, the type is inferred.
+    - **expression**: The initial value assigned to the variable or constant.
+    - **transfer** expression: Handles the assignment or movement of values, specifically relevant for resource types.
+
+- **Example**:
+  ```cadence
+  // Declare a constant named `a` of type Int
+  pub let a: Int = 10
+  
+  // Declare a variable named `b`, with type inference for the initial value
+  var b = 20
+  
+  // Reassign a new value to variable `b`
+  b = 30
+  ```
+
+- **Constants vs. Variables**:
+  - **Constants (let)**: Once a constant is initialized with a value, it cannot be reassigned. However, if the constant points to a mutable object (e.g., an array or resource), the object’s contents can be modified.
+    ```cadence
+    let a = 10
+    a = 15  // Invalid: Constants cannot be reassigned
+    ```
+  - **Variables (var)**: Variables allow for reassignment, meaning the identifier can be bound to a different value after initialization.
+    ```cadence
+    var b = 20
+    b = 30  // Valid: Variables can be reassigned
+    ```
+
+- **Initialization Requirement**:
+  - All variable declarations in Cadence must be initialized with a value at the time of declaration. It is invalid to declare a variable or constant without an initial value.
+    ```cadence
+   // Invalid: Declaring a constant without initialization
+   let x  // This will result in an error
+    ```
+
+- **Type Inference and Type Annotations**:
+  - Cadence allows for type inference, meaning that the type of a variable or constant can be automatically determined based on the initial value. However, you can also explicitly specify a type annotation to declare the type of the variable or constant.
+    ```cadence
+    // Type inference: Cadence infers that `a` is of type Int
+    let a = 10
+    
+    // Explicit type annotation: `b` is explicitly declared as Int
+    let b: Int = 20
+    ```
+
+- **Unique Identifiers**:
+  - Each identifier in a scope must be unique. Declaring another variable or constant with the same name within the same scope is not allowed, regardless of the declaration kind or type.
+    ```cadence
+    let x = 10
+    
+    // Invalid: Redeclaring a constant with the same name in the same scope
+    let x = 20  // This will result in an error
+    
+    var y = 5
+    
+    // Invalid: Redeclaring a variable with the same name in the same scope
+    var y = 10  // This will result in an error
+    ```
+
+- **Redeclaration in Sub-scopes**:
```

Comment:
Maybe reference scoping rules / and or move shadowing rules to scoping rules, given they are very closely related.